File: Python/credit_card_utils/credit_card_utils_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
AllToolkit - Credit Card Utilities Test Suite
==============================================
Comprehensive tests for credit_card_utils module.

Run with: python -m pytest credit_card_utils_test.py -v
Or directly: python credit_card_utils_test.py
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mod import (
    # Luhn functions
    luhn_checksum, validate_luhn,
    
    # Card type detection
    detect_card_type, get_card_type_name,
    
    # Card number operations
    clean_card_number, format_card_number, mask_card_number,
    validate_card_length,
    
    # CVV validation
    validate_cvv,
    
    # Expiry validation
    validate_expiry, parse_expiry_string,
    
    # Comprehensive validation
    get_card_info, validate_card,
    
    # Test card generation
    generate_test_card, generate_test_cards,
    
    # BIN/IIN lookup
    get_iin_info,
    
    # Utilities
    compare_cards, card_to_int, int_to_card,
    
    # Data classes
    CardInfo, ExpiryCheck,
    
    # Constants
    CARD_PATTERNS, CARD_TYPE_NAMES
)

import unittest
from datetime import datetime, timedelta


class TestLuhnAlgorithm(unittest.TestCase):
    """Test Luhn algorithm implementation."""
    
    def test_luhn_checksum(self):
        """Test Luhn checksum calculation."""
        # Known test cases: checksum for number without check digit
        self.assertEqual(luhn_checksum('411111111111111'), 1)  # Visa 4111111111111111
        self.assertEqual(luhn_checksum('424242424242424'), 2)  # Visa 4242424242424242
        self.assertEqual(luhn_checksum('555555555555444'), 4)  # MC 5555555555554444
    
    def test_validate_luhn_valid_cards(self):
        """Test Luhn validation with valid card numbers."""
        valid_cards = [
            '4111111111111111',  # Visa (known test card)
            '4242424242424242',  # Visa (known test card)
            '5555555555554444',  # Mastercard (known test card)
            '378282246310005',   # Amex (known test card)
            '371449635398431',   # Amex (known test card)
            '6011111111111117',  # Discover (known test card)
        ]
        for card in valid_cards:
            with self.subTest(card=card):
                self.assertTrue(validate_luhn(card))
    
    def test_validate_luhn_invalid_cards(self):
        """Test Luhn validation with invalid card numbers."""
        invalid_cards = [
            '4111111111111112',  # Wrong check digit
            '4242424242424243',  # Wrong check digit
            '5555555555554445',  # Wrong check digit
            '1234567890123456',  # Invalid
        ]
        for card in invalid_cards:
            with self.subTest(card=card):
                self.assertFalse(validate_luhn(card))
    
    def test_validate_luhn_with_formatting(self):
        """Test Luhn validation with formatted card numbers."""
        self.assertTrue(validate_luhn('4111-1111-1111-1111'))
        self.assertTrue(validate_luhn('4111 1111 1111 1111'))
        self.assertTrue(validate_luhn('  4111111111111111  '))
    
    def test_validate_luhn_edge_cases(self):
        """Test Luhn validation edge cases."""
        self.assertFalse(validate_luhn(''))
        self.assertFalse(validate_luhn('1'))
        self.assertFalse(validate_luhn('abc'))
        self.assertFalse(validate_luhn('1234abcd5678'))


class TestCardTypeDetection(unittest.TestCase):
    """Test card type detection."""
    
    def test_detect_visa(self):
        """Test Visa card detection."""
        visa_cards = [
            '4111111111111111',
            '4242424242424242',
            '4012888888881881',
        ]
        for card in visa_cards:
            card_type, info = detect_card_type(card)
            with self.subTest(card=card):
                self.assertEqual(card_type, 'visa')
    
    def test_detect_mastercard(self):
        """Test Mastercard detection."""
        mc_cards = [
            '5555555555554444',
            '2223000048400011',  # New Mastercard range
        ]
        for card in mc_cards:
            card_type, info = detect_card_type(card)
            with self.subTest(card=card):
                self.assertEqual(card_type, 'mastercard')
    
    def test_detect_amex(self):
        """Test American Express detection."""
        amex_cards = [
            '378282246310005',
            '371449635398431',
        ]
        for card in amex_cards:
            card_type, info = detect_card_type(card)
            with self.subTest(card=card):
                self.assertEqual(card_type, 'amex')
    
    def test_detect_discover(self):
        """Test Discover card detection."""
        discover_cards = [
            '6011111111111117',
            '6011000990139424',
        ]
        for card in discover_cards:
            card_type, info = detect_card_type(card)
            with self.subTest(card=card):
                self.assertEqual(card_type, 'discover')
    
    def test_detect_unknown(self):
        """Test unknown card type detection."""
        unknown_cards = [
            '9999999999999999',
            '1234567890123456',
        ]
        for card in unknown_cards:
            card_type, info = detect_card_type(card)
            with self.subTest(card=card):
                self.assertEqual(card_type, 'unknown')
    
    def test_get_card_type_name(self):
        """Test card type name retrieval."""
        self.assertEqual(get_card_type_name('4111111111111111'), 'Visa')
        self.assertEqual(get_card_type_name('5555555555554444'), 'Mastercard')
        self.assertEqual(get_card_type_name('378282246310005'), 'American Express')


class TestCardNumberOperations(unittest.TestCase):
    """Test card number cleaning and formatting."""
    
    def test_clean_card_number(self):
        """Test card number cleaning."""
        self.assertEqual(clean_card_number('4111-1111-1111-1111'), '4111111111111111')
        self.assertEqual(clean_card_number('4111 1111 1111 1111'), '4111111111111111')
        self.assertEqual(clean_card_number('  4111111111111111  '), '4111111111111111')
        self.assertEqual(clean_card_number('4111.1111.1111.1111'), '4111111111111111')
    
    def test_format_card_number_visa(self):
        """Test Visa card formatting."""
        self.assertEqual(format_card_number('4111111111111111'), '4111 1111 1111 1111')
    
    def test_format_card_number_amex(self):
        """Test Amex card formatting (4-6-5 format)."""
        self.assertEqual(format_card_number('378282246310005'), '3782 822463 10005')
    
    def test_mask_card_number(self):
        """Test card number masking."""
        masked = mask_card_number('4111111111111111')
        self.assertEqual(masked, '4111********1111')
        self.assertTrue(masked.startswith('4111'))
        self.assertTrue(masked.endswith('1111'))
    
    def test_mask_card_number_custom(self):
        """Test card number masking with custom options."""
        masked = mask_card_number('4111111111111111', mask_char='#', visible_start=2, visible_end=2)
        self.assertEqual(masked, '41############11')
    
    def test_mask_short_number(self):
        """Test masking of short numbers."""
        masked = mask_card_number('123456')
        self.assertEqual(masked, '123456')  # No mask for short numbers
    
    def test_validate_card_length(self):
        """Test card length validation."""
        # Valid Visa
        is_valid, lengths = validate_card_length('4111111111111111')
        self.assertTrue(is_valid)
        self.assertIn(16, lengths)
        
        # Invalid length for Visa
        is_valid, _ = validate_card_length('411111111111111')
        self.assertFalse(is_valid)
        
        # Valid Amex (15 digits)
        is_valid, lengths = validate_card_length('378282246310005')
        self.assertTrue(is_valid)
        self.assertIn(15, lengths)


class TestCVVValidation(unittest.TestCase):
    """Test CVV validation."""
    
    def test_valid_cvv_3_digit(self):
        """Test valid 3-digit CVV."""
        self.assertTrue(validate_cvv('123'))
        self.assertTrue(validate_cvv('999'))
        self.assertTrue(validate_cvv('000'))
    
    def test_valid_cvv_4_digit(self):
        """Test valid 4-digit CVV."""
        self.assertTrue(validate_cvv('1234'))
        self.assertTrue(validate_cvv('9999'))
    
    def test_invalid_cvv(self):
        """Test invalid CVV."""
        self.assertFalse(validate_cvv('12'))     # Too short
        self.assertFalse(validate_cvv('12345'))   # Too long
        self.assertFalse(validate_cvv('abc'))    # Non-numeric
        self.assertFalse(validate_cvv(''))       # Empty
    
    def test_cvv_for_card_type(self):
        """Test CVV validation with card type."""
        # Amex requires 4-digit CVV
        self.assertTrue(validate_cvv('1234', 'amex'))
        self.assertFalse(validate_cvv('123', 'amex'))
        
        # Visa uses 3-digit CVV
        self.assertTrue(validate_cvv('123', 'visa'))
        self.assertFalse(validate_cvv('1234', 'visa'))


class TestExpiryValidation(unittest.TestCase):
    """Test expiry date validation."""
    
    def test_valid_future_expiry(self):
        """Test valid future expiry date."""
        future_year = datetime.now().year + 1
        result = validate_expiry(12, future_year)
        self.assertTrue(result.is_valid)
        self.assertFalse(result.is_expired)
    
    def test_expired_date(self):
        """Test expired date detection."""
        past_year = datetime.now().year - 1
        result = validate_expiry(1, past_year)
        self.assertTrue(result.is_expired)
        self.assertFalse(result.is_valid)
    
    def test_invalid_month(self):
        """Test invalid month detection."""
        result = validate_expiry(13, datetime.now().year + 1)
        self.assertFalse(result.is_valid)
        self.assertIsNotNone(result.error)
    
    def test_expiry_formatting(self):
        """Test expiry date formatting."""
        result = validate_expiry(5, 2026)
        self.assertEqual(result.formatted, '05/26')
    
    def test_days_until_expiry(self):
        """Test days until expiry calculation."""
        future_year = datetime.now().year + 1
        result = validate_expiry(12, future_year)
        self.assertIsNotNone(result.days_until_expiry)
        self.assertGreater(result.days_until_expiry, 0)
    
    def test_parse_expiry_string(self):
        """Test expiry string parsing."""
        month, year = parse_expiry_string('12/25')
        self.assertEqual(month, 12)
        self.assertEqual(year, 2025)
        
        month, year = parse_expiry_string('06/2026')
        self.assertEqual(month, 6)
        self.assertEqual(year, 2026)
        
        month, year = parse_expiry_string('1225')
        self.assertEqual(month, 12)
        self.assertEqual(year, 2025)
    
    def test_parse_invalid_expiry(self):
        """Test parsing invalid expiry strings."""
        month, year = parse_expiry_string('invalid')
        self.assertIsNone(month)
        self.assertIsNone(year)


class TestComprehensiveValidation(unittest.TestCase):
    """Test comprehensive card validation."""
    
    def test_get_card_info(self):
        """Test comprehensive card info retrieval."""
        info = get_card_info('4111111111111111')
        
        self.assertIsInstance(info, CardInfo)
        self.assertEqual(info.card_type, 'visa')
        self.assertEqual(info.card_type_name, 'Visa')
        self.assertTrue(info.is_valid)
        self.assertTrue(info.is_valid_luhn)
        self.assertTrue(info.is_length_valid)
        self.assertEqual(info.cvv_length, 3)
    
    def test_get_card_info_amex(self):
        """Test Amex card info."""
        info = get_card_info('378282246310005')
        
        self.assertEqual(info.card_type, 'amex')
        self.assertEqual(info.cvv_length, 4)
        self.assertIn(15, info.expected_lengths)
    
    def test_validate_card_valid(self):
        """Test valid card validation."""
        is_valid, errors = validate_card('4111111111111111')
        self.assertTrue(is_valid)
        self.assertEqual(len(errors), 0)
    
    def test_validate_card_invalid_luhn(self):
        """Test invalid Luhn detection."""
        is_valid, errors = validate_card('4111111111111112')
        self.assertFalse(is_valid)
        self.assertIn('Luhn', errors[0])
    
    def test_validate_card_with_cvv(self):
        """Test card validation with CVV."""
        is_valid, _ = validate_card('4111111111111111', cvv='123')
        self.assertTrue(is_valid)
        
        is_valid, errors = validate_card('4111111111111111', cvv='12')
        self.assertFalse(is_valid)
    
    def test_validate_card_empty(self):
        """Test validation of empty card number."""
        is_valid, errors = validate_card('')
        self.assertFalse(is_valid)


class TestCardGeneration(unittest.TestCase):
    """Test card number generation."""
    
    def test_generate_visa(self):
        """Test Visa card generation."""
        card = generate_test_card('visa')
        self.assertTrue(validate_luhn(card))
        card_type, _ = detect_card_type(card)
        self.assertEqual(card_type, 'visa')
    
    def test_generate_mastercard(self):
        """Test Mastercard generation."""
        card = generate_test_card('mastercard')
        self.assertTrue(validate_luhn(card))
        card_type, _ = detect_card_type(card)
        self.assertEqual(card_type, 'mastercard')
    
    def test_generate_amex(self):
        """Test Amex generation."""
        card = generate_test_card('amex')
        self.assertTrue(validate_luhn(card))
        card_type, _ = detect_card_type(card)
        self.assertEqual(card_type, 'amex')
        self.assertEqual(len(card), 15)
    
    def test_generate_multiple(self):
        """Test generating multiple cards."""
        cards = generate_test_cards(10)
        self.assertEqual(len(cards), 10)
        for card in cards:
            self.assertTrue(validate_luhn(card))
    
    def test_generate_specific_length(self):
        """Test generating card with specific length."""
        card = generate_test_card('visa', length=16)
        self.assertEqual(len(card), 16)
    
    def test_generate_invalid_type(self):
        """Test generating with invalid card type."""
        with self.assertRaises(ValueError):
            generate_test_card('invalid_type')
    
    def test_generate_invalid_length(self):
        """Test generating with invalid length."""
        with self.assertRaises(ValueError):
            generate_test_card('visa', length=15)  # Visa doesn't support 15-digit


class TestIINLookup(unittest.TestCase):
    """Test BIN/IIN lookup."""
    
    def test_get_iin_info_visa(self):
        """Test Visa IIN lookup."""
        info = get_iin_info('411111')
        self.assertEqual(info['card_type'], 'visa')
        self.assertEqual(info['card_type_name'], 'Visa')
    
    def test_get_iin_info_amex(self):
        """Test Amex IIN lookup."""
        info = get_iin_info('378282')
        self.assertEqual(info['card_type'], 'amex')
    
    def test_get_iin_info_too_short(self):
        """Test IIN lookup with too short input."""
        info = get_iin_info('12345')
        self.assertIn('error', info)


class TestUtilities(unittest.TestCase):
    """Test utility functions."""
    
    def test_compare_cards(self):
        """Test card comparison."""
        self.assertEqual(compare_cards('4000000000000001', '4000000000000001'), 0)
        self.assertEqual(compare_cards('4000000000000001', '4000000000000002'), -1)
        self.assertEqual(compare_cards('4000000000000002', '4000000000000001'), 1)
    
    def test_card_to_int(self):
        """Test card to integer conversion."""
        self.assertEqual(card_to_int('4111111111111111'), 4111111111111111)
        self.assertEqual(card_to_int('4111-1111-1111-1111'), 4111111111111111)
    
    def test_int_to_card(self):
        """Test integer to card conversion."""
        self.assertEqual(int_to_card(4111111111111111), '4111111111111111')


class TestEdgeCases(unittest.TestCase):
    """Test edge cases and boundary conditions."""
    
    def test_empty_string(self):
        """Test handling of empty strings."""
        self.assertEqual(clean_card_number(''), '')
        self.assertEqual(card_to_int(''), 0)
    
    def test_non_digit_characters(self):
        """Test handling of non-digit characters."""
        self.assertEqual(clean_card_number('abcd'), '')
    
    def test_card_info_unknown_type(self):
        """Test card info for unknown type."""
        info = get_card_info('9999999999999999')
        self.assertEqual(info.card_type, 'unknown')
        self.assertEqual(info.card_type_name, 'Unknown')


if __name__ == '__main__':
    unittest.main(verbosity=2)

File: TypeScript/credit_card_utils/credit_card_utils_test.ts
/**
 * Credit Card Utilities Test Suite
 * 
 * Comprehensive tests for the credit card utilities module.
 * Run with: npx ts-node credit_card_utils_test.ts
 * Or compile: tsc credit_card_utils_test.ts && node credit_card_utils_test.js
 */

import {
  sanitize,
  format,
  mask,
  maskFormatted,
  validate,
  isValid,
  isValidFormat,
  isValidLength,
  luhnCheck,
  calculateLuhnCheckDigit,
  detectCardType,
  getCardInfo,
  getCardTypeName,
  getAllCardTypes,
  isCardTypeSupported,
  isValidCVV,
  validateExpiry,
  validateExpiryString,
  getBIN,
  isValidBIN,
  generateTestNumber,
  CardType,
  CreditCardUtils,
} from './mod';

// Test result tracking
let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`✓ ${name}`);
    passed++;
  } catch (error) {
    console.error(`✗ ${name}`);
    console.error(`  Error: ${error}`);
    failed++;
  }
}

function assertEqual(actual: any, expected: any, message?: string): void {
  if (actual !== expected) {
    throw new Error(`${message || 'Assertion failed'}: expected ${expected}, got ${actual}`);
  }
}

function assertTrue(value: boolean, message?: string): void {
  if (!value) {
    throw new Error(message || 'Expected true, got false');
  }
}

function assertFalse(value: boolean, message?: string): void {
  if (value) {
    throw new Error(message || 'Expected false, got true');
  }
}

function assertMatch(value: string, pattern: RegExp, message?: string): void {
  if (!pattern.test(value)) {
    throw new Error(`${message || 'Pattern match failed'}: ${value} does not match ${pattern}`);
  }
}

function assertThrows(fn: () => void, expectedMessage?: string): void {
  try {
    fn();
    // Function didn't throw, this is an error
    throw new Error(expectedMessage || 'Expected function to throw');
  } catch (e) {
    // Function threw something
    if (e instanceof Error) {
      // If this is our own "Expected function to throw" error, re-throw it
      if (e.message === (expectedMessage || 'Expected function to throw') && e.message === 'Expected function to throw') {
        throw e;
      }
      // Otherwise, the function threw as expected (possibly with a different message)
    }
    // Non-Error exceptions are also valid throws
  }
}

// ==================== Sanitization Tests ====================

test('sanitize() removes non-digits', () => {
  assertEqual(sanitize('4111-1111-1111-1111'), '4111111111111111');
  assertEqual(sanitize('4111 1111 1111 1111'), '4111111111111111');
  assertEqual(sanitize('4111.1111.1111.1111'), '4111111111111111');
  assertEqual(sanitize('abc4111111111111111def'), '4111111111111111');
});

test('sanitize() handles empty string', () => {
  assertEqual(sanitize(''), '');
  assertEqual(sanitize('   '), '');
  assertEqual(sanitize('abc'), '');
});

// ==================== Formatting Tests ====================

test('format() formats Visa card number', () => {
  assertEqual(format('4111111111111111'), '4111 1111 1111 1111');
  assertEqual(format('4111-1111-1111-1111'), '4111 1111 1111 1111');
});

test('format() formats Amex card number', () => {
  assertEqual(format('378282246310005'), '3782 822463 10005');
});

test('format() handles short numbers', () => {
  assertEqual(format('123'), '123');
  assertEqual(format('1234'), '1234');
  assertEqual(format('12345'), '1234 5');
});

// ==================== Masking Tests ====================

test('mask() masks card number', () => {
  assertEqual(mask('4111111111111111'), '4111********1111');
  assertEqual(mask('378282246310005'), '3782*******0005');
});

test('mask() with custom visible digits', () => {
  assertEqual(mask('4111111111111111', 6, 4), '411111******1111');
  assertEqual(mask('4111111111111111', 4, 2), '4111**********11');
});

test('mask() with custom mask character', () => {
  assertEqual(mask('4111111111111111', 4, 4, 'X'), '4111XXXXXXXX1111');
});

test('maskFormatted() masks with formatting', () => {
  const masked = maskFormatted('4111111111111111');
  // maskFormatted first masks then formats, so masked digits become the mask char
  // Expected: "4111" + "********" + "1111" -> formatted
  assertMatch(masked, /4111/);
  assertMatch(masked, /1111/);
});

// ==================== Luhn Check Tests ====================

test('luhnCheck() validates known test numbers', () => {
  // Valid test numbers
  assertTrue(luhnCheck('4111111111111111')); // Visa
  assertTrue(luhnCheck('4012888888881881')); // Visa
  assertTrue(luhnCheck('378282246310005')); // Amex
  assertTrue(luhnCheck('5555555555554444')); // MasterCard
  assertTrue(luhnCheck('5105105105105100')); // MasterCard
  assertTrue(luhnCheck('6011111111111117')); // Discover
});

test('luhnCheck() rejects invalid numbers', () => {
  assertFalse(luhnCheck('4111111111111112')); // Invalid check digit
  assertFalse(luhnCheck('1234567890123450')); // Truly invalid
  assertFalse(luhnCheck('5555555555554443')); // Invalid MasterCard
});

test('luhnCheck() handles empty and invalid input', () => {
  assertFalse(luhnCheck(''));
  assertFalse(luhnCheck('abcdef'));
  assertFalse(luhnCheck('123abc456'));
});

test('calculateLuhnCheckDigit() calculates correct check digits', () => {
  // Verify against known valid numbers
  // 4111111111111111 -> check digit for 411111111111111 is 1
  const visaPartial = '411111111111111';
  const visaCheck = calculateLuhnCheckDigit(visaPartial);
  assertTrue(luhnCheck(visaPartial + visaCheck));
  
  // 378282246310005 -> check digit for 37828224631000 is 5
  const amexPartial = '37828224631000';
  const amexCheck = calculateLuhnCheckDigit(amexPartial);
  assertTrue(luhnCheck(amexPartial + amexCheck));
  
  // Verify the full number is valid
  assertEqual(luhnCheck('4111111111111111'), true);
  assertEqual(luhnCheck('378282246310005'), true);
});

test('calculateLuhnCheckDigit() throws on invalid input', () => {
  assertThrows(() => calculateLuhnCheckDigit('abc'), 'Partial card number must contain only digits');
  assertThrows(() => calculateLuhnCheckDigit(''), 'Partial card number must contain only digits');
  assertThrows(() => calculateLuhnCheckDigit('123a456'), 'Partial card number must contain only digits');
});

// ==================== Card Type Detection Tests ====================

test('detectCardType() detects Visa', () => {
  assertEqual(detectCardType('4111111111111111'), CardType.VISA);
  assertEqual(detectCardType('4012888888881881'), CardType.VISA);
  assertEqual(detectCardType('4222222222222'), CardType.VISA);
});

test('detectCardType() detects MasterCard', () => {
  assertEqual(detectCardType('5555555555554444'), CardType.MASTERCARD);
  assertEqual(detectCardType('5105105105105100'), CardType.MASTERCARD);
  assertEqual(detectCardType('2221000000000009'), CardType.MASTERCARD);
  assertEqual(detectCardType('2720999999999995'), CardType.MASTERCARD);
});

test('detectCardType() detects Amex', () => {
  assertEqual(detectCardType('378282246310005'), CardType.AMEX);
  assertEqual(detectCardType('371449635398431'), CardType.AMEX);
});

test('detectCardType() detects Discover', () => {
  assertEqual(detectCardType('6011111111111117'), CardType.DISCOVER);
  assertEqual(detectCardType('6011000990139424'), CardType.DISCOVER);
  assertEqual(detectCardType('6511111111111111'), CardType.DISCOVER);
});

test('detectCardType() detects JCB', () => {
  assertEqual(detectCardType('3530111333300000'), CardType.JCB);
  assertEqual(detectCardType('3566002020360505'), CardType.JCB);
});

test('detectCardType() detects Diners Club', () => {
  assertEqual(detectCardType('30569309025904'), CardType.DINERS_CLUB);
  assertEqual(detectCardType('36000000000008'), CardType.DINERS_CLUB);
});

test('detectCardType() detects UnionPay', () => {
  assertEqual(detectCardType('6212345678901234'), CardType.UNIONPAY);
  assertEqual(detectCardType('6225881234567890'), CardType.UNIONPAY);
});

test('detectCardType() detects Mir', () => {
  assertEqual(detectCardType('2200123456789012'), CardType.MIR);
  assertEqual(detectCardType('2204123456789016'), CardType.MIR);
});

test('detectCardType() detects Maestro', () => {
  assertEqual(detectCardType('5018123456789012'), CardType.MAESTRO);
  assertEqual(detectCardType('5612345678901234'), CardType.MAESTRO);
});

test('detectCardType() returns UNKNOWN for invalid input', () => {
  assertEqual(detectCardType(''), CardType.UNKNOWN);
  assertEqual(detectCardType('123'), CardType.UNKNOWN);
  assertEqual(detectCardType('0000000000000000'), CardType.UNKNOWN);
});

// ==================== Validation Tests ====================

test('validate() validates correct card numbers', () => {
  const result = validate('4111111111111111');
  assertTrue(result.valid);
  assertEqual(result.cardType, CardType.VISA);
  assertEqual(result.error, null);
  assertTrue(result.info !== null);
});

test('validate() rejects empty input', () => {
  const result = validate('');
  assertFalse(result.valid);
  assertEqual(result.error, 'Card number is empty');
});

test('validate() rejects invalid characters in short input', () => {
  const result = validate('abc123'); // After sanitize: '123' (too short)
  assertFalse(result.valid);
  assertMatch(result.error!, /length is invalid|empty/);
});

test('validate() rejects wrong length', () => {
  const result = validate('123');
  assertFalse(result.valid);
  assertMatch(result.error!, /length is invalid/);
});

test('validate() rejects invalid Luhn', () => {
  const result = validate('4111111111111112');
  assertFalse(result.valid);
  assertEqual(result.error, 'Card number fails Luhn check');
});

test('validate() rejects wrong length for card type', () => {
  const result = validate('3782822463100050'); // Amex with 16 digits
  assertFalse(result.valid);
  assertMatch(result.error!, /Invalid length for American Express/);
});

test('isValid() returns boolean', () => {
  assertTrue(isValid('4111111111111111'));
  assertFalse(isValid('4111111111111112'));
  assertFalse(isValid(''));
});

test('isValidFormat() checks format', () => {
  assertTrue(isValidFormat('4111111111111111'));
  assertTrue(isValidFormat('123456789012'));
  assertFalse(isValidFormat('123'));
  assertFalse(isValidFormat('12345678901234567890123'));
});

test('isValidLength() checks length for card type', () => {
  assertTrue(isValidLength('4111111111111111')); // Visa 16 digits (valid)
  assertTrue(isValidLength('378282246310005')); // Amex 15 digits (valid)
  assertTrue(isValidLength('4111111111111')); // Visa 13 digits (valid - Visa supports 13, 16, 19)
  assertFalse(isValidLength('411111111111')); // Visa 12 digits (invalid)
  assertFalse(isValidLength('37828224631000')); // Amex 14 digits (invalid - Amex only 15)
});

// ==================== CVV Validation Tests ====================

test('isValidCVV() validates 3-digit CVV', () => {
  assertTrue(isValidCVV('123'));
  assertTrue(isValidCVV('999'));
  assertTrue(isValidCVV('000'));
  assertFalse(isValidCVV('12')); // Too short
  assertFalse(isValidCVV('1')); // Too short
  assertFalse(isValidCVV('abc')); // Invalid chars
  assertFalse(isValidCVV('12a')); // Invalid chars
});

test('isValidCVV() validates 4-digit CVV for Amex', () => {
  assertTrue(isValidCVV('1234', CardType.AMEX));
  assertTrue(isValidCVV('123', CardType.VISA));
  assertFalse(isValidCVV('123', CardType.AMEX));
  assertFalse(isValidCVV('1234', CardType.VISA));
});

// ==================== Expiry Date Tests ====================

test('validateExpiry() validates correct dates', () => {
  const now = new Date();
  const futureYear = now.getFullYear() + 1;
  const result = validateExpiry(12, futureYear);
  assertTrue(result.valid);
  assertEqual(result.month, 12);
  assertEqual(result.year, futureYear);
  assertFalse(result.expired);
});

test('validateExpiry() handles 2-digit years', () => {
  const result = validateExpiry(12, 30);
  assertTrue(result.valid);
  assertEqual(result.year, 2030);
});

test('validateExpiry() rejects invalid months', () => {
  const result = validateExpiry(13, 2030);
  assertFalse(result.valid);
  assertMatch(result.error!, /Invalid month/);
  
  const result2 = validateExpiry(0, 2030);
  assertFalse(result2.valid);
});

test('validateExpiry() detects expired cards', () => {
  const result = validateExpiry(1, 2020);
  assertTrue(result.valid);
  assertTrue(result.expired);
});

test('validateExpiry() detects cards expiring soon', () => {
  const now = new Date();
  const currentMonth = now.getMonth() + 1;
  const currentYear = now.getFullYear();
  
  // Test with current month/year (should expire soon)
  const result = validateExpiry(currentMonth, currentYear);
  assertTrue(result.valid);
  assertTrue(result.expiresSoon || result.expired);
});

test('validateExpiryString() parses MM/YY format', () => {
  const result = validateExpiryString('12/30');
  assertTrue(result.valid);
  assertEqual(result.month, 12);
  assertEqual(result.year, 2030);
});

test('validateExpiryString() parses MM/YYYY format', () => {
  const result = validateExpiryString('12/2030');
  assertTrue(result.valid);
  assertEqual(result.month, 12);
  assertEqual(result.year, 2030);
});

test('validateExpiryString() handles spaces', () => {
  const result = validateExpiryString('12 / 30');
  assertTrue(result.valid);
  assertEqual(result.month, 12);
  assertEqual(result.year, 2030);
});

test('validateExpiryString() rejects invalid formats', () => {
  const result = validateExpiryString('1230');
  assertFalse(result.valid);
  
  const result2 = validateExpiryString('12-30');
  assertTrue(result2.valid); // Should accept dash separator
});

// ==================== BIN Tests ====================

test('getBIN() extracts BIN', () => {
  assertEqual(getBIN('4111111111111111'), '411111');
  assertEqual(getBIN('4111111111111111', 8), '41111111');
  assertEqual(getBIN('123', 6), null);
});

test('isValidBIN() validates BIN format', () => {
  assertTrue(isValidBIN('411111')); // 6 digits - valid
  assertTrue(isValidBIN('41111111')); // 8 digits - valid
  assertTrue(isValidBIN('4111111')); // 7 digits - also valid (6-8 range)
  assertFalse(isValidBIN('41111')); // 5 digits - invalid
  assertFalse(isValidBIN('411111111')); // 9 digits - invalid
  assertFalse(isValidBIN('abcdef')); // Non-digits - invalid
});

// ==================== Card Info Tests ====================

test('getCardInfo() returns card info', () => {
  const info = getCardInfo(CardType.VISA);
  assertTrue(info !== null);
  assertEqual(info!.type, CardType.VISA);
  assertEqual(info!.displayName, 'Visa');
  assertEqual(info!.cvvLength, 3);
});

test('getCardInfo() returns null for unknown', () => {
  const info = getCardInfo(CardType.UNKNOWN);
  assertEqual(info, null);
});

test('getAllCardTypes() returns all supported types', () => {
  const types = getAllCardTypes();
  assertTrue(types.length > 0);
  assertTrue(types.every(t => t.type !== CardType.UNKNOWN));
});

test('isCardTypeSupported() checks support', () => {
  assertTrue(isCardTypeSupported('VISA'));
  assertTrue(isCardTypeSupported('visa'));
  assertTrue(isCardTypeSupported('MASTERCARD'));
  assertFalse(isCardTypeSupported('UNKNOWN'));
  assertFalse(isCardTypeSupported('FAKE'));
});

test('getCardTypeName() returns display name', () => {
  assertEqual(getCardTypeName('4111111111111111'), 'Visa');
  assertEqual(getCardTypeName('378282246310005'), 'American Express');
  assertEqual(getCardTypeName('5555555555554444'), 'MasterCard');
});

// ==================== Test Number Generation Tests ====================

test('generateTestNumber() generates valid Visa number', () => {
  const number = generateTestNumber(CardType.VISA);
  assertTrue(luhnCheck(number));
  assertEqual(detectCardType(number), CardType.VISA);
  // Visa supports 13, 16, 19 - default to 16
  assertTrue(number.length === 16 || number.length === 13 || number.length === 19);
});

test('generateTestNumber() generates valid Amex number', () => {
  const number = generateTestNumber(CardType.AMEX);
  assertEqual(number.length, 15); // Amex is always 15 digits
  assertTrue(luhnCheck(number));
  assertEqual(detectCardType(number), CardType.AMEX);
});

test('generateTestNumber() generates valid MasterCard number', () => {
  const number = generateTestNumber(CardType.MASTERCARD);
  assertEqual(number.length, 16); // MasterCard is 16 digits
  assertTrue(luhnCheck(number));
  assertEqual(detectCardType(number), CardType.MASTERCARD);
});

test('generateTestNumber() throws for unknown type', () => {
  assertThrows(() => generateTestNumber(CardType.UNKNOWN));
});

// ==================== Namespace Tests ====================

test('CreditCardUtils namespace exports all functions', () => {
  assertTrue(typeof CreditCardUtils.sanitize === 'function');
  assertTrue(typeof CreditCardUtils.format === 'function');
  assertTrue(typeof CreditCardUtils.mask === 'function');
  assertTrue(typeof CreditCardUtils.validate === 'function');
  assertTrue(typeof CreditCardUtils.luhnCheck === 'function');
  assertTrue(typeof CreditCardUtils.detectCardType === 'function');
  assertTrue(typeof CreditCardUtils.CardType === 'object');
});

// ==================== Edge Cases ====================

test('handles formatted input throughout', () => {
  const formatted = '4111 1111 1111 1111';
  assertTrue(isValid(formatted));
  assertEqual(detectCardType(formatted), CardType.VISA);
  assertEqual(sanitize(formatted), '4111111111111111');
});

test('handles long card numbers', () => {
  const longVisa = '4111111111111111111'; // 19 digits
  assertEqual(detectCardType(longVisa), CardType.VISA);
});

test('handles short card numbers', () => {
  assertEqual(detectCardType('4'), CardType.VISA); // Still detected by prefix
  assertFalse(isValidLength('4'));
});

// ==================== Run Summary ====================

console.log('\n' + '='.repeat(50));
console.log(`Test Results: ${passed} passed, ${failed} failed`);
console.log('='.repeat(50));

if (failed > 0) {
  process.exit(1);
}